docs/arcaflow/workflows/expressions.md

```diff
@@ -28,31 +85,274 @@ $.foo.bar
 
 ## Dot notation
 
-The dot notation allows you to dive into a data structure. If the data structure is a map, it will look up the map keys. If it's a list, it will look up the list item. If it's an object, it will look up the object property.
+The dot notation allows you to reference fields of an object.
 
-For example, you can look up the list item from this data structure:
+For example, if you have an object on the root data structure named "a" with the field "b" in it, you can access it with:
 
+```
+$.a.b
+```
+
+## Bracket accessor
+
+The bracket accessor is used for referencing values in maps or lists.
+
+
+#### List access
+For list access, you specify the index of the value you want to access. 
+Lists are zero-indexed (so the first value has an index of 0).
+
+If you have a list named `foo`:
 
 ```yaml
 foo:
   - Hello world!
 ```
 
-You can use the following expression to access it:
+You can access the first value with the expression:
+```JavaScript
+$.foo[0]
+```
+
+Giving the output `"Hello world!"`
+
+#### Map access
+
+Maps, also known as dictionaries in some languages, are key-value pair data structures.
 
+To use a map in an expression, the expression to the left of the brackets must be a reference to a map. That is then followed by a pair of brackets with a sub-expression between them. That sub-expression must evaluate to a valid key in the map.
+
+Here is an example of a map whose key are strings, and whose values are integers. The map is stored in a field called `foo` in the root-level object:
```

Comment:
"key" should be plural, "keys are strings". Or maybe just "map with string keys and integer values" rather than falling into the semantic abyss of assigning ownership to inanimates.